lib/Dialect/FIRRTL/Transforms/LowerTypes.cpp

```diff
@@ -252,6 +254,123 @@ void FIRRTLTypesLowering::visitDecl(InstanceOp op) {
   opsToRemove.push_back(op);
 }
 
+void FIRRTLTypesLowering::visitDecl(MemOp op) {
+
+  auto type = op.getDataTypeOrNull();
+  // Bail out if the memory is empty
+  if (!type)
+    return;
+
+  SmallVector<FlatBundleFieldEntry, 8> fieldTypes;
+  flattenBundleTypes(op.getDataTypeOrNull(), "", false, fieldTypes);
+
+  auto maybeFlip = [](Type a) -> std::tuple<BundleType, bool> {
+    return TypeSwitch<Type, std::tuple<BundleType, bool>>(a)
+        .Case<FlipType>([](FlipType b) {
+          return std::make_tuple(b.getElementType().cast<BundleType>(), true);
+        })
+        .Default([](Type b) {
+          return std::make_tuple(b.cast<BundleType>(), false);
+        });
+  };
+
+  DenseMap<StringRef, Value> newWires;
+  for (auto field : fieldTypes) {
+    // Create the new type for the memory. Copy every field, but
+    // update the "data" field to be the groundn type and the "mask"
+    // field to be a uint<1> type.
+    SmallVector<Type, 8> resultTypes;
+    for (size_t i = 0, e = op.getNumResults(); i != e; ++i) {
+      SmallVector<BundleType::BundleElement, 5> resultFields;
+
+      BundleType portType;
+      bool isFlipped;
+      std::tie(portType, isFlipped) = maybeFlip(op.getResult(i).getType());
```

Comment:
This might be a little too meta... not sure if it would be easier to just write this logic out imperatively.

Reply:
Agreed.  You can express higher order and functional things in C++, but it really isn't designed to be used that way.

Reply:
Sounds good. Find if I use a struct for the return type of `maybeFlip` here?

Reply:
Switched to a struct in a fixup commit.

Reply:
I think using a tuple would be better than a struct, but I'm guessing they are looking for something more like:
```
auto isFlipped = op.getResult(i).isa<FlipType>();
auto underlyingType = op.getResult(i).getPassiveType().cast<BundleType>();
```

Reply:
😬 Good suggestion @youngar. Switched to that in a fixup.